### PR TITLE
NOISSUE - Add opc-ua type handling and unsubscription

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

### CODE REVIEW

File: opcua/gopcua/browser.go [hunk 1]
```diff
@@ -82,7 +85,7 @@ func (c browser) Browse(serverURI, nodeID string) ([]opcua.BrowsedNode, error) {
 }
 
 func browse(n *opcuaGopcua.Node, path string, level int) ([]NodeDef, error) {
-	if level > 10 {
+	if level > maxChildrens {
 		return nil, nil
```
Comment (mteodor) [inline opcua/gopcua/browser.go:89]: no error here?

Comment (manuio) [reply opcua/gopcua/browser.go:89]: No need, it's the end of the iteration.

File: opcua/gopcua/subscribe.go [hunk 2]
```diff
@@ -99,15 +108,15 @@ func (c client) Subscribe(cfg opcua.Config) error {
...
+	if err := c.runHandler(sub, cfg.ServerURI, cfg.NodeID); err != nil {
+		c.logger.Warn(fmt.Sprintf("Unsubscribed from OPC-UA node %s.%s: %s", cfg.ServerURI, cfg.NodeID, err))
 	}
 
 	return nil
 }
 
-func (c client) runHandler(sub *opcuaGopcua.Subscription, cfg opcua.Config) error {
-	nodeID, err := uaGopcua.ParseNodeID(cfg.NodeID)
+func (c client) runHandler(sub *opcuaGopcua.Subscription, uri, node string) error {
```
Comment (mteodor) [inline opcua/gopcua/subscribe.go:118]: this method maybe should be split in smaller methods

File: opcua/gopcua/subscribe.go [hunk 3]
```diff
@@ -145,32 +154,43 @@ func (c client) runHandler(sub *opcuaGopcua.Subscription, cfg opcua.Config) erro
...
+						msg.Data = item.Value.Value.Time()
 					default:
 						msg.Data = 0
 					}
 
 					if err := c.publish(token, msg); err != nil {
-						c.logger.Warn(fmt.Sprintf("failed to publish: %s", err))
+						switch err {
+						case errNotFoundServerURI, errNotFoundNodeID, errNotFoundConn:
+							return err
```
Comment (mteodor) [inline opcua/gopcua/subscribe.go:190]: I dont understand how is this error returned?
you are waiting on the channel

Comment (manuio) [reply opcua/gopcua/subscribe.go:190]: This function runs in a go routine that will stop if there is an error with the routemap

File: opcua/gopcua/subscribe.go (hunk 2 above)
Comment (manuio) [reply opcua/gopcua/subscribe.go:118]: I agree, it's just that for now I'm trying to keep this code similar to an example that I'm using

**Review by mteodor — APPROVED** [review]

File: opcua/gopcua/subscribe.go (hunk 2 above)
Comment (mteodor) [reply opcua/gopcua/subscribe.go:118]: so you call it here https://github.com/manuio/mainflux/blob/opcua/opcua/gopcua/subscribe.go#L111
and it stays in runHandler until error occurs?

Comment (manuio) [reply opcua/gopcua/subscribe.go:118]: yes

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**